modules/drivers/snowflake/src/metabase/driver/snowflake.clj

```diff
@@ -834,52 +834,92 @@
+  (-> (m/remove-vals str/blank? row)
+      (m/update-existing :data_type normalize-type)
+      (update :autoincrement boolean)
+      (assoc-database-required)
+      (dissoc :kind :null? :database_name)
+      (set/rename-keys {:schema_name :table-schema
+                        :data_type :database-type
+                        :table_name :table-name
+                        :column_name :name
+                        :comment :field-comment
+                        :autoincrement :database-is-auto-increment})))
+
+(defmethod sql-jdbc.sync/describe-fields-pre-process-xf :snowflake
+  [driver db & {:keys [schema-names table-names] :as _args}]
+  (let [schema-names (set schema-names)
+        table-names (set table-names)
+        position-counter (volatile! {})
+        positioner (map (fn [{:keys [schema-name table-name] :as row}]
+                          (let [idx [schema-name table-name]
+                                pos (inc (get @position-counter idx -1))]
+                            (vswap! position-counter assoc idx pos)
+                            (assoc row :database-position pos))))
+        normalize-type (memoize normalize-type*)
+        pks (sql-jdbc.execute/do-with-connection-with-options
+             driver db nil
              (fn [^java.sql.Connection conn]
-               (with-open [stmt (.prepareStatement conn (format "show primary keys in database \"%s\";" (get-in database [:details :db])))
+               (with-open [stmt (.prepareStatement conn (format "show primary keys in database \"%s\";" (get-in db [:details :db])))
                            rset (.executeQuery stmt)]
-                 (into #{} (map (juxt :schema_name :table_name :column_name)) (resultset-seq rset)))))]
-    (eduction
-     (map (fn [col]
-            (let [lookup ((juxt :table-schema :table-name :name) col)
-                  pk? (contains? pks lookup)]
-              (assoc col :pk? pk?))))
-     (apply (get-method driver/describe-fields :sql-jdbc) driver database args))))
+                 (into #{} (map (juxt :schema_name :table_name :column_name)) (resultset-seq rset)))))
+        normalize-row (comp
+                       (remove #(= (:schema_name %) "INFORMATION_SCHEMA"))
+                       (map #(normalize-describe-fields-row % normalize-type))
+                       positioner
+                       (map (fn [col]
+                              (let [lookup ((juxt :table-schema :table-name :name) col)
+                                    pk? (contains? pks lookup)]
+                                (assoc col :pk? pk?))))
+                       (transducers/sorted-by (juxt :table-schema :table-name :database-position)))]
+    (cond->> normalize-row
+      ;; Add pre-filter to schemas and tables (schemas are checked first)
+      (seq table-names)
+      (comp (filter #(contains? table-names (:table_name %))))
+
+      (seq schema-names)
+      (comp (filter #(contains? schema-names (:schema_name %)))))))
```

Comment:
I suggest making the comment even more prominent, similar to how the QP middleware collection is documented. 

Maybe even inline `normalize-row` so that there a single thread building the transducer, even if that means most conditions will be `true`.  Then you could use `cond->` and have the transducers in the natural order.